controllers/k8ssandra/cleanup.go

```diff
@@ -206,154 +361,64 @@ func (r *K8ssandraClusterReconciler) deleteDc(ctx context.Context, kc *api.K8ssa
 		// There is no need to requeue here. Reconciliation will be trigger by updates made by cass-operator.
 		return result.Done()
 	}
-
-	delete(kc.Status.Datacenters, dcName)
-	logger.Info("DC deletion finished", "DC", dcName)
-	return result.Continue()
+	return nil
 }
 
-func (r *K8ssandraClusterReconciler) findStargateForDeletion(
-	ctx context.Context,
-	kcKey client.ObjectKey,
-	dcName string,
-	remoteClient client.Client) (*stargateapi.Stargate, client.Client, error) {
-
+func (r *K8ssandraClusterReconciler) findRemoteClientForReaper(ctx context.Context, kcKey client.ObjectKey, dcName string) (client.Client, error) {
```

Comment:
We know the remote client is going to be the same for all components.

In the original code, as soon as we found a client for Stargate or Reaper, it would get passed to the following methods instead of looking it up again. I think we should preserve that in some form.

In fact we can even improve the lookup, because the current approach is very inefficient. We don't need to look for the components to delete in every context, we already know the context because it's declared in the DC's `K8sContext` field.
The only problem is that we only have the DC name right now, but that's an easy fix:
- in `checkDcDeletion()`, either modify `GetDatacenterForDecommission` so that it returns the full `CassandraDatacenterTemplate`, or re-iterate `kc.Spec.Cassandra.Datacenters` to look it up by name.
- then look up the client:
    ```go
    remoteClient, err := r.ClientCache.GetRemoteClient(dcTemplate.K8sContext)
    if err != nil { ... }
    ```
- and finally pass it into `deleteDc` and its child methods.

That should allow us to remove all the `findRemoteClient` methods.

Reply:
Ok, I had trouble figuring out this bit, and the old code was confusing. But this makes sense now.

Reply:
> The only problem is that we only have the DC name right now, but that's an easy fix:
> 
> * in `checkDcDeletion()`, either modify `GetDatacenterForDecommission` so that it returns the full `CassandraDatacenterTemplate`, or re-iterate `kc.Spec.Cassandra.Datacenters` to look it up by name.

Actually it's not that easy, because the DC has already been removed from `kc.Spec.Cassandra.Datacenters` by the time we start the cleanup.